cmd/imagePushToRegistry.go

```diff
@@ -0,0 +1,235 @@
+	err := runImagePushToRegistry(&config, telemetryData, utils)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+
+func runImagePushToRegistry(config *imagePushToRegistryOptions, telemetryData *telemetry.CustomData, utils imagePushToRegistryUtils) error {
+	if len(config.TargetImages) == 0 {
+		config.TargetImages = config.SourceImages
+	}
+
+	if len(config.TargetImages) != len(config.SourceImages) {
+		log.SetErrorCategory(log.ErrorConfiguration)
+		return errors.New("configuration error: please configure targetImage and sourceImage properly")
+	}
+
+	re := regexp.MustCompile(`^https?://`)
+	config.SourceRegistryURL = re.ReplaceAllString(config.SourceRegistryURL, "")
+	config.TargetRegistryURL = re.ReplaceAllString(config.TargetRegistryURL, "")
+
+	log.Entry().Debug("Handling source registry credentials")
+	if err := handleCredentialsForPrivateRegistry(config.DockerConfigJSON, config.SourceRegistryURL, config.SourceRegistryUser, config.SourceRegistryPassword, utils); err != nil {
+		return errors.Wrap(err, "failed to handle credentials for source registry")
+	}
+
+	log.Entry().Debug("Handling destination registry credentials")
+	if err := handleCredentialsForPrivateRegistry(config.DockerConfigJSON, config.TargetRegistryURL, config.TargetRegistryUser, config.TargetRegistryPassword, utils); err != nil {
+		return errors.Wrap(err, "failed to handle credentials for target registry")
+	}
+
+	if len(config.LocalDockerImagePath) > 0 {
+		if err := pushLocalImageToTargetRegistry(config, utils); err != nil {
+			return errors.Wrapf(err, "failed to push local image to %q", config.TargetRegistryURL)
+		}
+		return nil
+	}
+
+	if err := copyImages(config, utils); err != nil {
+		return errors.Wrap(err, "failed to copy images")
+	}
+
+	return nil
+}
+
+func handleCredentialsForPrivateRegistry(dockerConfigJsonPath, registry, username, password string, utils imagePushToRegistryUtils) error {
+	if len(dockerConfigJsonPath) == 0 && (len(registry) == 0 || len(username) == 0 || len(password) == 0) {
+		return errors.New("docker credentials not provided")
+	}
+
+	if len(dockerConfigJsonPath) == 0 {
+		if _, err := docker.CreateDockerConfigJSON(registry, username, password, "", targetDockerConfigPath, utils); err != nil {
+			return errors.Wrap(err, "failed to create new docker config")
+		}
+		return nil
+	}
+
+	if _, err := docker.CreateDockerConfigJSON(registry, username, password, targetDockerConfigPath, dockerConfigJsonPath, utils); err != nil {
+		return errors.Wrapf(err, "failed to update docker config %q", dockerConfigJsonPath)
+	}
+
+	if err := docker.MergeDockerConfigJSON(targetDockerConfigPath, dockerConfigJsonPath, utils); err != nil {
+		return errors.Wrapf(err, "failed to merge docker config files")
+	}
+
+	return nil
+}
+
+func copyImages(config *imagePushToRegistryOptions, utils imagePushToRegistryUtils) error {
+	g, ctx := errgroup.WithContext(context.Background())
```

Comment:
should make concurrency limit configurable?
we can use this pattern: https://go.dev/play/p/HovNRgp6FxH

I don't know if there any limits from Artifactory about max parallel uploads and so on.

Reply:
Answered [here](https://github.com/SAP/jenkins-library/pull/4609#discussion_r1409533559)